File: 0108-convert-sorted-array-to-binary-search-tree/0108-convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        return self.addNode(nums, 0, len(nums) - 1)
    
    def addNode(self, nums: List[int], start, end) -> Optional[TreeNode]:
        if start > end:
            return None

        mid = start + (end - start) // 2
        currentNode = TreeNode(nums[mid])
        currentNode.left = self.addNode(nums, start, mid - 1)
        currentNode.right = self.addNode(nums, mid + 1, end)

        return currentNode
        

File: 0108-convert-sorted-array-to-binary-search-tree/0108-convert-sorted-array-to-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        int numsSize = nums.size();
        TreeNode* root = nullptr;
        addNode(nums, 0, numsSize - 1, root);

        return root;
    }
private:
    void addNode(vector<int>& nums, int start, int end, TreeNode*& currentNode) {
        if (start > end) {
            return;
        }

        int mid = start + (end - start) / 2;
        currentNode = new TreeNode(nums[mid]);

        addNode(nums, start, mid - 1, currentNode->left);
        addNode(nums, mid + 1, end, currentNode->right);
    }
};